subprojects/jvm-services/src/integTest/groovy/org/gradle/jvm/toolchain/InvalidJvmInstallationReportingIntegrationTest.groovy

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import spock.lang.IgnoreIf
+import spock.lang.TempDir
+
+import java.util.regex.Pattern
+
+class InvalidJvmInstallationReportingIntegrationTest extends AbstractIntegrationSpec {
+
+    @IgnoreIf({ GradleContextualExecuter.isNoDaemon() || GradleContextualExecuter.isConfigCache() || AvailableJavaHomes.differentJdk == null })
+    def "invalid JDK is cached only for current build if in daemon"() {
+        // Require a different JDK to be able to find the logs of its probing for system properties
+        def existingJdk = AvailableJavaHomes.differentJdk
+
+        def invalidJdkHome1 = new File(temporaryDirectory, "jdk1")
+        def invalidJdkHome2 = new File(temporaryDirectory, "jdk2")
+        assert(invalidJdkHome1.mkdirs())
+        assert(invalidJdkHome2.mkdirs())
+
+        // The builds should trigger toolchains discovery; here it's done from different subprojects in order to test
+        // that the JVM installation detection is cached during a build
+        multiProjectBuild("invalidJvmInstallationReporting", ["sub1", "sub2"])
+        buildFile << """
+            import org.gradle.jvm.toolchain.internal.JavaInstallationRegistry;
+
+            subprojects {
```

Comment:
While convenient, I would instead use a single block applied to both build scripts since we want to get rid of that construct in the future.